### DISCUSSION TITLE
What's the webpack 5 story for web workers written in typescript?

### DISCUSSION THREAD

**Discussion by azangru** [webpack/webpack#13655] [Q&A]
Hi,

The documentation for webpack 5 [says](https://webpack.js.org/guides/web-workers/) that webpack no longer needs `worker-loader` to load web worker files.

This is reinforced by the README on the `worker-loader` page, which warns that `worker-loader` is only intended for use with webpack 4 and is not supposed to be used with webpack 5.

However, the example of worker usage in the webpack docs is very simple — it assumes that developers author their worker files in javascript. How should one configure webpack to be able to load workers authored in typescript?

**Comment by alexander-akait** [comment]
https://www.typescriptlang.org/play?#code/HYUw7gBA6g9gTgaxHAFKSBVASgGRQcgDoB6MeJOQgKwGd8AaCASwFsAHeAF0JZE4ENCAVzgAbAJTiA3EA, what is the problem?

**Reply by azangru** [reply]
I mean, it wouldn't be `worker.js` anymore :smiley: 

I think the reason for my confusion was that I didn't understand that the syntax `new URL('./worker', import.meta.url)`, which looks like vanilla browser API, will actually be processed by webpack as a URL asset, as described [here](https://webpack.js.org/guides/asset-modules/#url-assets). I only figured it out yesterday after I posted this question here. I was assuming that the developer who wanted to load a web worker had first to somehow find out where the final destination path of that worker.js file will be, and pass that to the URL constructor. I wonder if you were expecting your users, when reading the web workers chapter, to already know about webpack's URL assets. For me, it was a novelty that didn't click immediately, and it would have been very helpful if the documentation on web workers referred to the section on URL assets to explain why this syntax works.

**Reply by alexander-akait** [reply]
@azangru feel free to open an issue in our docs https://github.com/webpack/webpack.js.org, we will improve this

> I was assuming that the developer who wanted to load a web worker had first to somehow find out where the final destination path of that worker.js file will be, and pass that to the URL constructor.

webpack does this job for you, just use `new URL('./worker.js', import.meta.url)` and put your code in `worker.js`, doesn't forget about `js` extension, it is URL

**Reply by azangru** [reply]
Thanks.

I have a follow-up question though. The documentation for worker-loader talks about how workers are restricted by the same-site policy and thus have to be loaded from the same host. But in my setup (custom server-side-rendered React), I have, when developing locally, a rendering server which serves html documents on localhost:8080, and a server for static assets on localhost:8081, on which webpack is running. What are my options in webpack 5 for loading web workers with this setup?

**Reply by alexander-akait** [reply]
I think here two solutions:
- proxy
- use `target: 'webworker'` and use multi compiler mode https://github.com/webpack/webpack/tree/main/examples/multi-compiler

**Reply by azangru** [reply]
What does the `target: 'webworker'` do? Does it compile the worker so that it gets requested as a blob, similar to what worker-loader's `inline: fallback` option did, or is it doing something different? The docs are very quiet on this subject.

**Reply by alexander-akait** [reply]
It generate file for webworker env, i.e. you can use it in `new Worker` (without `import.meta.url`), we don't support `inline`, because it was workaround for old browsers, but it can be solved custom loader or we can implement `blob-loader`, because you can use blob not only for web workers

**Reply by azangru** [reply]
I tried to load the worker using the URL assets syntax, but can't get past the errors.

My worker file is written in typescript and is located at the following path: `src/shared/workers/sequenceFetcher.worker.ts`. All my typescript files are processed with babel-loader, if that's relevant:

```
    module: {
      rules: [
        {
          test: /.tsx?$/,
          loader: 'babel-loader',
          exclude: /node_modules/
        },
    }
```

If I request my worker file using its typescript extension, or if I omit the extension entirely, like with regular typescript imports:

```
  const workerUrl = new URL('src/shared/workers/sequenceFetcher.worker.ts', import.meta.url);
```

the url that webpack resolves contains a stringified Object: `http://localhost:8080/[object%20Module]`.

Whereas if I request it using a .js extension:

```
  const workerUrl = new URL('src/shared/workers/sequenceFetcher.worker.js', import.meta.url);
```

webpack, quite rightly, tells me that the file `sequenceFetcher.worker.js` doesn't exist (because of course it doesn't; this file has a ts extension).

So I am lost. How do i transform `http://localhost:8080/[object%20Module]` into a path to the file?

**Reply by alexander-akait** [reply]
Why do you use `new URL` for workers url? Please use `new Worker('./foo.js', import.meta.url)`, but we can improve this and allow to use `new Worker(new URL('./foo.js', import.meta.url))`

**Reply by azangru** [reply]
Sorry, I am confused 😞 Could we please go step by step slowly here?

> Why do you use new URL for workers url? 

Because this is what the [docs](https://webpack.js.org/guides/web-workers/#root) are currently telling me to do:

![image](https://user-images.githubusercontent.com/6834224/123962306-395c9000-d9a9-11eb-818b-4ba63bd3b00b.png)

Are the docs incorrect?

> Please use new Worker('./foo.js', import.meta.url)

- Will webpack be able to intercept this and to change `./foo.js` to the actual destination public path of the foo file?
- If my worker source file is `./foo.ts` rather than `./foo.js` (typescript rather than javascript), should I write `new Worker('./foo.ts', import.meta.url)`? Will this work? If not, what's the correct way of importing web workers whose source is written in typescript.

**Reply by alexander-akait** [reply]
> Will webpack be able to intercept this and to change ./foo.js to the actual destination public path of the foo file?

Should not worry about it. I want to use do not use `publicPath` if you don't use CDN, in webpack v5 it is auto and all urls are relative in assets

I said a little wrong, you should use `new Worker(new URL('./foo.js', import.meta.url))`, but you can't use (it will work, but `worker.js` will handle as asset, not worker):
```
const myWorkerURL = new URL('./foo.js', import.meta.url);
const worker = new Worker(myWorkerURL);
```

**Reply by azangru** [reply]
> I said a little wrong, you should use new Worker(new URL('./foo.js', import.meta.url)),

Right. Which brings me back to my previous problem. When I try to load my worker file, which, remember, is written in typescript:

```
new Worker(new URL('src/shared/workers/sequenceFetcher.worker.ts', import.meta.url));
```

the url that is produced from this is `http://localhost:8080/[object%20Module]`. I suspect this is the case because webpack applies babel-loader to all my typescript files and does not transform this specific file into an asset that has an asset url. Is there a way to get from `http://localhost:8080/[object%20Module]` to `http://localhost:8080/my_actual_public_path/worker_file_name.js`?

**Reply by alexander-akait** [reply]
Something wrong with configuration, can your provide?

**Reply by azangru** [reply]
Ok, can we start with this minimal repro case please?

https://github.com/azangru/web-worker-ts-webpack-test

I've tried to replicate my use case. Unfortunately for me, this minimal repro case is not showing the exact same error with `[object%20Module]` as I am getting in my real project. However, even in this minimal example, there still is an error. Although webpack seems to correctly transform the worker file from typescript into javascript, the url of the worker file that webpack generates ends in `.ts` (see screenshot below), and the browser doesn't know how to handle this:

![image](https://user-images.githubusercontent.com/6834224/124043538-1e1e6e80-da03-11eb-9227-f9cf631aa2ae.png)

**Reply by azangru** [reply]
Regarding my minimal repro case in the message above. I updated the `assetModuleFilename` field in the webpack config to manually rewrite the `.ts` extension with `.js`:

```
  output: {
    filename: 'main.js',
    path: path.resolve(__dirname, 'dist'),
    assetModuleFilename: (pathData) => {
      const { filename } = pathData;

      if (filename.endsWith('.ts')) {
        return '[name]-[hash].js'
      } else {
        return '[hash][ext][query]';
      }
    }
  },
```

It almost works. My next immediate problem is that I am importing a dependency in the worker file; and it seems I cannot do that, because the browser is now telling me:

![image](https://user-images.githubusercontent.com/6834224/124047941-224f8980-da0d-11eb-985a-120ce0a418f9.png)

It was so much easier with the worker loader!

**Reply by alexander-akait** [reply]
Look at here https://github.com/azangru/web-worker-ts-webpack-test/pull/1/, using `new URL` outside `new Worker` creates asset not web worker, because it is statically analyzable, also **always** use extension at the end of `import`, because it is URL

**Reply by azangru** [reply]
Yay, fantastic, thanks so much! At least something is working :-) Now I need to get back to my real code and figure out what was creating the `[object%20Module]` in the url. I'll probably be back 😃 

There were so many teaching moments here. I'll try to submit an issue to webpack/webpack.js.org to improve the docs.

**Reply by alexander-akait** [reply]
Yes, feel free to feedback

**Reply by harveyconnor** [reply]
@azangru have you had anymore success? We're having trouble implementing ts web workers with our application using webpack 5. A repo of something or a tutorial of some sorts would be greatly appreciated!

**Reply by DustinJSilk** [reply]
@alexander-akait  is it possible to get this working with node worker threads?
I've tried importing worker threads and node URL in the example repo above, however, it stops compiling the TS and only copies across the raw ts file.

**Reply by DustinJSilk** [reply]
I've updated the example repository showing worker threads not compiling: https://github.com/DustinJSilk/web-worker-ts-webpack-test

**Reply by DustinJSilk** [reply]
Right I've managed to get it to compile with a little hack. I've added "DOM" to my tsconfig `lib`, and then casting the WebWorker URL to a Node URL seems to work.

```JavaScript
import { Worker } from "worker_threads";
import { URL as NodeURL } from "url";

const worker = new Worker(
  new URL("./workers/foo.worker.ts", import.meta.url) as NodeURL
);
```

**Reply by azangru** [reply]
@harveyconnor yes, web workers are working fine for me in a typescript codebase with webpack 5. I still have a repo with minimal repro case up, with @alexander-akait 's commit that made it work. You can check it out here: https://github.com/azangru/web-worker-ts-webpack-test

**Reply by alexander-akait** [reply]
@DustinJSilk do you solve the problem?

**Reply by kpratik2015** [reply]
> Look at here [azangru/web-worker-ts-webpack-test#1](https://github.com/azangru/web-worker-ts-webpack-test/pull/1), using `new URL` outside `new Worker` creates asset not web worker, because it is statically analyzable, also **always** use extension at the end of `import`, because it is URL

Hi, is there a workaround for this as I am facing this issue when making my web worker work cross-origin:

```ts
export const webWorker = new Worker(getWorkerUrl(new URL('./worker.ts', import.meta.url)));

function getWorkerURL(url: URL) {
 const content = `importScripts("${url}");`;
 return URL.createObjectURL(new Blob([content], {type: 'text/javascript'}))
}
```

**Reply by vankop** [reply]
@kpratik2015 your code should work cross-origin or I did not understand your problem...

**Reply by kpratik2015** [reply]
@vankop Problem is that with this code snippet, webpack is statically analyzing and creating asset.
Above only works if I use craco to override webpack config with following:
```js
output: {
  assetModuleFilename: (pathData) => {
      const { filename } = pathData;
  
      if (filename.endsWith('.ts')) {
          return '[name].js';
      } else {
          return '[name][ext]';
      }
  },
}
```

**Reply by vankop** [reply]
yeah, this creates an asset, to achieve what you want you need to use multicompiler mode, example webpack.config:
```js
module.exports = [
  {
     name: "webworker",
     target: "webworker",
     entry: "./worker.ts",
     output: { filename: "worker.js" }
  },
  {
     name: "app",
     dependencies: ["worker"]
  }
];
```
and then use in `new URL(...)` output of the first compilation

**Reply by kpratik2015** [reply]
Ah, okay. I was hoping for a way without modifying webpack config. Thanks!

**Reply by subhamX** [reply]
This is such a nice discussion. Thank you folks, finally I was able to properly configure web worker using typescript + webpack5 with async/await. 🙂

**Reply by mulgurul** [reply]
Ok @subhamX but how did you manage it. Please share an example:-)
Thanks

**Reply by subhamX** [reply]
@mulgurul; Just saw this man. I completely forgot what I did. It was during my time at mercari (proprietary code, and I don't have access to it anymore).

But maybe, I used this multicompiler thing to compile the worker first, and then the app.

And while using the worker in app, I initialized it globally, instead of using `useRef` or `createRef` (for some reason, the singleton pattern using these hooks didn't work for me).

**Comment by jmcdl** [comment]
Hi all, I've been going back and forth over this thread for an hour but haven't found a solution to my very similar problem here:

I'm creating the worker as so:
![image](https://user-images.githubusercontent.com/25539870/131734292-92f101f9-eac8-45de-821e-7bfbfbacaafe.png)

The webpack terminal output shows the worker file created as _src_workers_project_worker_ts.19dc6e0a66c2986d1a57.js_:
![image](https://user-images.githubusercontent.com/25539870/131734490-3c8274cc-7d2f-4823-a14d-d44ef3969358.png)

However the browser console shows that the app is trying to load a worker file with an undefined contentHash:
![image](https://user-images.githubusercontent.com/25539870/131734671-bb846071-df67-494d-9e75-d3481bfef742.png)

Is anyone able to suggest how I might resolve this?

Thanks

**Reply by alexander-akait** [reply]
Fix here https://github.com/webpack/webpack/pull/14112

**Reply by jmcdl** [reply]
Thank you! I'll keep an eye out for the release.

**Comment by absurdprofit** [comment]
Hey guys there seems to be an issue in my situation. So I have a project which I bundle for use in a React project. When I try to use ```new Worker(new URL('./reader.worker.ts', import.meta.url));``` it compiles and outputs a separate file called ```src_reader_worker_ts.js``` that seems like a duplicate of the main file but when I use the separated syntax:
```
const workerURL = new URL('./reader.worker.ts', import.meta.url);
const worker = new Worker(workerURL);
```
It compiles fine with just the compiled js contents of the typescript worker file. However neither work and they both cause an ```Uncaught SyntaxError: Unexpected token '<' (at reader.worker.js:1:1)```
Because the import is wrong somehow and the React app returns HTML instead.

Both React App and project use webpack 5. The project webpack config:
```
module.exports = {
    mode: 'development',
    entry: {
        index: path.resolve(__dirname, 'src', 'index.ts'),
    },
    module: {
        rules: [
            {
                test: /\.ts?$/,
                use: 'ts-loader',
                exclude: [
                    /node_modules/,
                    /examples/
                ]
            }
        ]
    },
    resolve: {
        extensions: ['.ts', '...']
    },
    output: {
        path: path.resolve(__dirname, 'build'),
        libraryTarget: 'commonjs',
        globalObject: 'this',
        umdNamedDefine: true,
        assetModuleFilename: (pathData) => {
            const { filename } = pathData;

            if (filename.endsWith('.ts')) {
                return '[name].js';
            } else {
                return '[name][ext]';
            }
        },
        clean: true
    },
    devtool: 'source-map'
};
```

**Reply by vankop** [reply]
> Uncaught SyntaxError: Unexpected token '<' (at reader.worker.js:1:1)

this means you setup your webserver wrong way ( or there is no `reader.worker.js`  and server fallbacks to 404 or index html )

>  So I have a project which I bundle for use in a React project. When I try to use new Worker(new URL('./reader.worker.ts', import.meta.url)); it compiles and outputs a separate file called src_reader_worker_ts.js that seems like a duplicate of the main file

that is correct behaviour, maybe your webworker code is wrong..

**Reply by absurdprofit** [reply]
I didn't configure the webserver. It's the default that comes with create-react-app and invoked with `npm start`. The webworker is apart of a npm package so when I import modules from the package it all works fine but the worker import just doesn't work.

**Reply by mulgurul** [reply]
I read here: [https://webpack.js.org/guides/web-workers/](https://webpack.js.org/guides/web-workers/) that you should not declare URL as variable. Do the definition inside of the Worker constructor call ie: new Worker(new URL('./reader.worker.ts', import.meta.url)):

**Comment by Aloento** [comment]
I think HMR will conflict with this.
In top of  my generated file for `worker.ts`:
`__webpack_require__.$Refresh$.runtime = require('path/node_modules/react-refresh/runtime.js');`

then I will get  
`Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'Promise')` from react-refresh

But in fact Promise can be used in web worker with modern browsers and I have tested that the async can use in my browser.
So I don't know why it  said undefined.

**Comment by mulgurul** [comment]
I have spent more tha a day trying to make this work with electron, typescript, commonjs (no import.meta.url) and webpack 5 way. Been through so many hacks suggested by copilot and my own humble intelligence:-), but none really worked!
It would be so great if someone with knowledge about how this should be setup could bring the light in. Eventually. the Webpack documentation could be improved with some nice examples for further help.
Today I'm going to try if the worker-loader from v4 days and see if it still works.

**Reply by alexander-akait** [reply]
@mulgurul Just provide reproducible example and I will show you, then we can improve our docs

**Reply by mulgurul** [reply]
Great you will help Alexander..
Yes, heres a clean example, our app is based on the ElectronReactBoilerplate from github. Heres a clone with webworker added with different approaches inside app.tsx.
[https://github.com/mulgurul/ErbWithWebWorker](https://github.com/mulgurul/ErbWithWebWorker) 

Nb when the error is something with "<" it is probably because fetching the script falls through to index.html as default, and it starts with `<!DOCTYPE html>`

Thanks so much for looking into it. I'm quite frustrated by loosing many hours of work....

Best regards from Peter

**Reply by mulgurul** [reply]
@alexander-akait are you maybe able to take a look at this and see if theres a way to solve it? Thanks.

**Reply by alexander-akait** [reply]
You can use:
```
const worker = new Worker(
  // @ts-ignore
  new URL('./webworker.worker.ts', import.meta.url),
);
```

webpack will make it works after bundling, but yeah, ts doesn't allow to use `import.meta.url` with `"module": "commonjs",`, as alternative you cna use multi compiler mode and use `target: "webworker"` to bundle worker code before the whole application
